mirrord/cli/src/wizard.rs

```diff
@@ -0,0 +1,668 @@
+            }
+            TargetType::StatefulSet => {
+                seeker
+                    .list_all_namespaced::<StatefulSet>(None, None)
+                    .filter_map(|x| into_info(x, &seeker, &client))
+                    .collect::<Vec<_>>()
+                    .await
+            }
+            TargetType::Service => {
+                seeker
+                    .list_all_namespaced::<Service>(None, None)
+                    .filter_map(|x| into_info(x, &seeker, &client))
+                    .collect::<Vec<_>>()
+                    .await
+            }
+            TargetType::ReplicaSet => {
+                seeker
+                    .list_all_namespaced::<ReplicaSet>(None, None)
+                    .filter_map(|x| into_info(x, &seeker, &client))
+                    .collect::<Vec<_>>()
+                    .await
+            }
+            TargetType::Targetless => vec![], // the frontend does not yet support targetless
+        });
+    }
+
+    Ok(serde_json::to_string(&targets)?)
+}
+
+/// A helper function that bridges the gap between the types in
+/// [`list_all_namespaced`][seeker::KubeResourceSeeker::list_all_namespaced] and
+/// [`into_info`][IntoTargetInfo::into_info]. Emits `warn!`s when `into_info` produces an error.
+async fn into_info<T: IntoTargetInfo>(
+    target_type: kube::Result<T>,
+    seeker: &KubeResourceSeeker<'_>,
+    client: &Client,
+) -> Option<TargetInfo> {
+    target_type
+        .ok()?
+        .into_info(seeker, client)
+        .await
+        .unwrap_or_else(|error| {
+            tracing::warn!("{error}");
+            None
+        })
+        .map(
+            |TargetInfo {
+                 target_path,
+                 target_namespace,
+                 detected_ports,
+             }| TargetInfo {
+                target_path,
+                target_namespace,
+                detected_ports: detected_ports.into_iter().unique().collect(),
+            },
+        )
+}
+
+/// Implementable by resources that can be targeted by mirrord and can produce a [`TargetInfo`].
+trait IntoTargetInfo {
+    /// Construct an instance of [`TargetInfo`] from a valid target Resource. Fetches further
+    /// resources in some cases, to detect ports on the resource's pods.
+    ///
+    /// If fetching further resources fails, return a `CliError`. If there are any empty fields in
+    /// the Resource spec, return `None`.
+    fn into_info(
+        self,
+        seeker: &KubeResourceSeeker,
+        client: &Client,
+    ) -> impl Future<Output = Result<Option<TargetInfo>, CliError>>;
+}
+
+impl IntoTargetInfo for Pod {
+    async fn into_info(
+        self,
+        _seeker: &KubeResourceSeeker<'_>,
+        _client: &Client,
+    ) -> Result<Option<TargetInfo>, CliError> {
+        fn into_info_option(pod: Pod) -> Option<TargetInfo> {
+            let target_name = pod.metadata.name.clone()?;
+            let target_namespace = pod.metadata.namespace.clone()?;
```

Comment:
nit: I think there's a `name()` and `namespace()` methods on `Resource` types, so you don't have to access `metadata` in here.

P.S. applies to all the other targets.